macrosynergy/download/jpmaqs.py

```diff
@@ -24,616 +24,387 @@
+        if data_df.empty:
+            return False
+        return True
+
+    def check_connection(self, verbose: bool = False) -> bool:
+        """Check if the interface is connected to the server.
+        :return <bool>: True if connected, False if not.
         """
+        return self.dq_interface.check_connection(verbose=verbose)
 
-        tickers_no = len(_dict.keys())
-        length = list(_dict.values())[0].shape[0]
-
-        arr = np.empty(shape=(length * tickers_no, 3 + no_metrics), dtype=object)
-
-        i = 0
-        for k, v in _dict.items():
-            ticker = k.split("_")
-
-            cid = ticker[0]
-            xcat = "_".join(ticker[1:])
-
-            cid_broad = np.repeat(cid, repeats=v.shape[0])
-            xcat_broad = np.repeat(xcat, repeats=v.shape[0])
-            data = np.column_stack((cid_broad, xcat_broad, v))
-
-            row = i * v.shape[0]
-            arr[row : row + v.shape[0], :] = data
-            i += 1
-
-        columns = ["cid", "xcat", "real_date"]
-        cols_output = columns + original_metrics
-
-        df = pd.DataFrame(data=arr, columns=cols_output)
-
-        df["real_date"] = pd.to_datetime(df["real_date"], yearfirst=True)
-        df = df[df["real_date"].dt.dayofweek < 5]
-        df = df.fillna(value=np.nan)
-        df = df.reset_index(drop=True)
-
-        for m in original_metrics:
-            df[m] = df[m].astype(dtype=np.float32)
-
-        df.real_date = pd.to_datetime(df.real_date)
-        return df
-
-    def check_connection(self) -> Tuple[bool, dict]:
-        """
-        Interface to the DQ API's check_connection method,
-        which in turn checks the connection (heartbeat) to the DQ API.
+    def validate_download_args(
+        self,
+        tickers: List[str],
+        cids: List[str],
+        xcats: List[str],
+        metrics: List[str],
+        start_date: str,
+        end_date: str,
+        expressions: List[str],
+        show_progress: bool,
+        as_dataframe: bool,
+    ) -> bool:
+        """Validate the arguments passed to the download function.
+
+        :params -- see macrosynergy.download.jpmaqs.JPMaQSDownload.download()
+
+        :return <bool>: True if valid.
+
+        :raises <ValueError>: if provided arguments are invalid or
+            semantically incorrect.
         """
 
-        with dataquery.Interface(**self.dq_args) as dq:
-            return dq.check_connection()
+        def is_valid_date(date: str) -> bool:
+            try:
+                datetime.datetime.strptime(date, "%Y-%m-%d")
+                return True
+            except ValueError:
+                return False
 
-    @staticmethod
-    def remove_jpmaqs_expr_formatting(expressions: List[str]) -> List[Tuple[str, str]]:
-        """
-        Removes the DB(JPMAQS, <ticker>, <metric>) formatting from a list of JPMaQS expressions.
+        try:
+            assert (
+                isinstance(tickers, list) or tickers is None
```

Comment:
we do. but it would still have to be passed as a list like : ['expression']. I'll add functionality to do the same for tickers, and expressions; but enforce cids and xcats to be passed as lists.

Reply:
This has been changed. Expressions, tickers, xcats, cids are all cast to list if a string is passed